core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -103,34 +113,27 @@ public void onComplete() {
             topicPartitionData, shareFetchData.groupId(), shareFetchData.fetchParams());
 
         try {
-            Seq<Tuple2<TopicIdPartition, LogReadResult>> responseLogResult = replicaManager.readFromLog(
-                shareFetchData.fetchParams(),
-                CollectionConverters.asScala(
-                    topicPartitionData.entrySet().stream().map(entry ->
-                        new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
-                ),
-                QuotaFactory.UNBOUNDED_QUOTA,
-                true);
-
-            Map<TopicIdPartition, FetchPartitionData> responseData = new HashMap<>();
-            responseLogResult.foreach(tpLogResult -> {
-                TopicIdPartition topicIdPartition = tpLogResult._1();
-                LogReadResult logResult = tpLogResult._2();
-                FetchPartitionData fetchPartitionData = logResult.toFetchPartitionData(false);
-                responseData.put(topicIdPartition, fetchPartitionData);
-                return BoxedUnit.UNIT;
-            });
-
-            log.trace("Data successfully retrieved by replica manager: {}", responseData);
+            Map<TopicIdPartition, LogReadResult> responseData;
+            if (partitionsAlreadyFetched.isEmpty())
+                responseData = readFromLog(topicPartitionData);
+            else
+                // There can't be a case when we have a non-empty logReadResponse, and we have a fresh topicPartitionData
+                // using tryComplete because purgatory ensures that 2 tryCompletes calls do not happen at the same time.
```

Comment:
I understand there can't be concurrent calls to tryComplete but I didn't get this comment. Though forceComplete() cannot be called twice. But forceComplete() on expiration can be on different thread and tryComplete() on different can this change any behaviour here?

Reply:
What I mean is that `partitionsAlreadyFetched` value can't be changed once we enter this point in `onComplete` via `forceComplete` either by expiration or by a `tryComplete` successful call. IIUC, `forceComplete` uses this `AtomicBoolean` variable `completed` which is used as locking mechanism for `forceComplete`. This should ensure atomicity of global variables between `tryComplete` and `forceComplete` we use in `DelayedShareFetch`